server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java

```diff
@@ -0,0 +1,194 @@
+            EditPLResp editPLResp,
+            DBObjectType plType) throws InterruptedException {
+        String plName = editPLReq.getPlName();
+        String editPLSql = editPLReq.getSql();
+        String tempPLSql = editPLSql.replaceFirst(plName, ODC_TEMPORARY_PROCEDURE);
+        SyncJdbcExecutor syncJdbcExecutor = connectionSession.getSyncJdbcExecutor(
+                ConnectionSessionConstants.CONSOLE_DS_KEY);
+        Lock editPLLock = obtainEditPLLock(connectionSession, plType);
+        if (!editPLLock.tryLock(OBTAIN_LOCK_TIME, TimeUnit.SECONDS)) {
+            throw new ConflictException(ErrorCodes.ResourceModifying, "Can not acquire jdbc lock");
+        }
+        try {
+            syncJdbcExecutor.execute(tempPLSql);
+            String dropTempPLSql = "drop " + plType + " if exists " + ODC_TEMPORARY_PROCEDURE;
+            syncJdbcExecutor.execute(dropTempPLSql);
+            String dropPLSql = "drop " + plType + " if exists " + plName;
+            syncJdbcExecutor.execute(dropPLSql);
+            syncJdbcExecutor.execute(editPLSql);
+            return editPLResp;
+        } finally {
+            editPLLock.unlock();
+        }
+    }
+
+    private Lock obtainEditPLLock(@NotNull ConnectionSession connectionSession, @NonNull DBObjectType plType) {
+        ConnectionConfig connConfig =
+                (ConnectionConfig) ConnectionSessionUtil.getConnectionConfig(connectionSession);
+        Long dataSourceId = connConfig.getId();
+        String lockKeyString;
+        switch (plType) {
+            case PROCEDURE:
+            case FUNCTION:
+            case TRIGGER:
+                lockKeyString = String.format("%s-%d", plType, dataSourceId);
+                break;
+            default:
+                throw new IllegalArgumentException(
+                        String.format("Unsupported pl type %s for dataSourceId %d", plType, dataSourceId));
+        }
+        return jdbcLockRegistry.obtain(lockKeyString);
+    }
+
+    private EditPLResp adaptSqlIntercept(@NotNull EditPLReq editPLReq, @NotNull ConnectionSession connectionSession,
+            boolean needSqlRuleCheck) {
+        SqlAsyncExecuteReq sqlAsyncExecuteReq = new SqlAsyncExecuteReq();
+        sqlAsyncExecuteReq.setSql(editPLReq.getSql());
+        sqlAsyncExecuteReq.setSplit(false);
+        List<OffsetString> sqls = sqlAsyncExecuteReq.ifSplitSqls()
+                ? SqlUtils.splitWithOffset(connectionSession, sqlAsyncExecuteReq.getSql(),
+                        sessionProperties.isOracleRemoveCommentPrefix())
+                : Collections.singletonList(new OffsetString(0, sqlAsyncExecuteReq.getSql()));
+        List<SqlTuple> sqlTuples = connectConsoleService.generateSqlTuple(sqls, connectionSession, sqlAsyncExecuteReq);
+        SqlAsyncExecuteResp response = SqlAsyncExecuteResp.newSqlAsyncExecuteResp(sqlTuples);
+        Map<String, Object> context = new HashMap<>();
+        context.put(SqlCheckInterceptor.NEED_SQL_CHECK_KEY, needSqlRuleCheck);
```

Comment:
these are all sql execution module's inner details, you may not care about these